src/math-land/invert.py

```diff
@@ -0,0 +1,106 @@
+from typing import Self
+
+
+class PrimaryColor:
+    """Base class for each primary color in the RGB space
+
+    >>> c = PrimaryColor(100)
+    >>> print(c)
+    100
+
+    >>> c.another_attr = 200
+    Traceback (most recent call last):
+        ...
+    AttributeError: 'PrimaryColor' object has no attribute 'another_attr'
+    """
+
+    __slots__ = ("value",)
+
+    def __init__(self, value: int) -> None:
+        """
+        Inititalize and check if the PrimaryColor has a valid value
+
+        >>> PrimaryColor(230.5)
+        Traceback (most recent call last):
+            ...
+        TypeError: Color value must be an int
+
+        >>> PrimaryColor(888)
+        Traceback (most recent call last):
+            ...
+        ValueError: Color value is not between 0 and 255
+        """
+        if not isinstance(value, int):
+            raise TypeError("Color value must be an int")
+
+        if not 0 <= value <= 255:
+            raise ValueError("Color value is not between 0 and 255")
+
+        self.value = value
+
+    def __invert__(self) -> Self:
+        """
+        Returns the inverse of the color as the complement to 255
+
+        >>> print(~PrimaryColor(100))
+        155
+        """
+        return PrimaryColor(255 - self.value)
+
+    def __str__(self) -> str:
+        return str(self.value)
+
+
+class RGBColor:
+    """
+    A color in RGB space.
+
+    >>> pink = RGBColor(255, 192, 203)
+    >>> print(pink)
+    (255, 192, 203)
+
+    """
+
+    __slots__ = "r", "g", "b"
+
+    def __init__(
+        self,
+        r: int | PrimaryColor,
+        g: int | PrimaryColor,
+        b: int | PrimaryColor,
+    ) -> None:
+        self.r = PrimaryColor(r) if isinstance(r, int) else r
+        self.g = PrimaryColor(g) if isinstance(g, int) else g
+        self.b = PrimaryColor(b) if isinstance(b, int) else b
```

Comment:
```suggestion
        self.r = r if isinstance(r, PrimaryColor) else PrimaryColor(r)
        self.g = g if isinstance(g, PrimaryColor) else PrimaryColor(g)
        self.b = b if isinstance(b, PrimaryColor) else PrimaryColor(b)
```

This makes it so that if a value of an incompatible type is passed, it goes through the checks in `PrimaryColor` and fails there.

I'd actually suggest making `PrimaryColor(PrimaryColor(2))` function so that this entire section could be just `self.r = PrimaryColor(r)` for each one, and it'd work similarly to some Python types like `int(int(5.3))`.

Reply:
> This makes it so that if a value of an incompatible type is passed, it goes through the checks in PrimaryColor and fails there.

I don't see how this is achieved?

> I'd actually suggest making `PrimaryColor(PrimaryColor(2))` function so that this entire section could be just `self.r = PrimaryColor(r)` for each one, and it'd work similarly to some Python types like `int(int(5.3))`.

"making it idempotent" is the term you're looking for :)

Reply:
Oh yeah that's right, I thought we were considering an "always `int` or `PrimaryColor`" case (should've made that clearer maybe 😅).

Reply:
IMO it does make sense to believe on the type-hint, but as I said on another comment, the code on `PrimaryColor` doesn't believe on the type-hint, and checks the type, so it'd be inconsistent IMO to have `PrimaryColor` care and `RGBColor` not care.

Reply:
That was definitively a test I was missing!